File: ts-client/src/authorization.ts
import { Principal } from '@dfinity/principal';
import { SigningKey, computeAddress, getBytes, hexlify, keccak256 } from 'ethers';

const encoder = new TextEncoder();
const decoder = new TextDecoder();

export type Address = Uint8Array;

export function deriveAddress(privateKey: Uint8Array): Address {
  if (privateKey.length !== 32) {
    throw new Error('private key must be 32 bytes');
  }
  const privateKeyHex = hexlify(privateKey);
  const addressHex = computeAddress(privateKeyHex);
  return hexToBytes(addressHex);
}

export function authorizationMessage(
  canisterId: Principal,
  address: Address,
  transportPublicKey: Uint8Array,
  expiryNs: bigint,
  nonce: bigint,
): Uint8Array {
  const message = `ICP Stealth Authorization:\naddress: 0x${bytesToHex(address)}\ncanister: ${canisterId.toText()}\ntransport: 0x${bytesToHex(transportPublicKey)}\nexpiry_ns:${expiryNs}\nnonce:${nonce}`;
  return eip191Message(encoder.encode(message));
}

export function signAuthorization(message: Uint8Array, privateKey: Uint8Array): Uint8Array {
  if (message.length === 0) {
    throw new Error('message must not be empty');
  }
  if (privateKey.length !== 32) {
    throw new Error('private key must be 32 bytes');
  }
  const digestHex = keccak256(message);
  const signingKey = new SigningKey(hexlify(privateKey));
  const signature = signingKey.sign(digestHex);
  return getBytes(signature.serialized);
}

export function unixTimeNs(): bigint {
  const nowMs = BigInt(Date.now());
  return nowMs * 1_000_000n;
}

export function bytesToHex(bytes: Uint8Array): string {
  return Array.from(bytes, (b) => b.toString(16).padStart(2, '0')).join('');
}

export function hexToBytes(hex: string): Uint8Array {
  const normalized = hex.startsWith('0x') ? hex.slice(2) : hex;
  if (normalized.length % 2 !== 0) {
    throw new Error('hex string must have even length');
  }
  const bytes = new Uint8Array(normalized.length / 2);
  for (let i = 0; i < bytes.length; i++) {
    bytes[i] = parseInt(normalized.slice(i * 2, i * 2 + 2), 16);
  }
  return bytes;
}

function eip191Message(message: Uint8Array): Uint8Array {
  const prefix = `\x19Ethereum Signed Message:\n${message.length}`;
  const prefixBytes = encoder.encode(prefix);
  const result = new Uint8Array(prefixBytes.length + message.length);
  result.set(prefixBytes, 0);
  result.set(message, prefixBytes.length);
  return result;
}

export function addressToHex(address: Address): string {
  return `0x${bytesToHex(address)}`;
}

export function messageToString(message: Uint8Array): string {
  return decoder.decode(message);
}

File: canisters/key_manager/src/authorization.rs
use alloy_primitives::{utils::eip191_message, Signature as AlloySignature, B256};
use candid::Principal;
use sha3::{Digest, Keccak256};
use thiserror::Error;

pub type Address = [u8; 20];

#[derive(Debug, Error)]
pub enum AuthorizationError {
    #[error("signature must be 65 bytes (r||s||v)")]
    InvalidLength,
    #[error("invalid signature")]
    InvalidSignature,
    #[error("failed to recover signer")]
    RecoveryFailed,
}

pub fn authorization_message(
    canister_id: Principal,
    address: &Address,
    transport_public_key: &[u8],
    expiry_ns: u64,
    nonce: u64,
) -> Vec<u8> {
    let message = format!(
        "ICP Stealth Authorization:\naddress: 0x{}\ncanister: {}\ntransport: 0x{}\nexpiry_ns:{expiry_ns}\nnonce:{nonce}",
        hex::encode(address),
        canister_id.to_text(),
        hex::encode(transport_public_key),
    );
    eip191_message(message.as_bytes())
}

pub fn recover_address_from_signature(
    message: &[u8],
    signature: &[u8],
) -> Result<Address, AuthorizationError> {
    let signature =
        AlloySignature::from_raw(signature).map_err(|_| AuthorizationError::InvalidSignature)?;
    let digest = Keccak256::digest(message);
    let mut hash_bytes = [0u8; 32];
    hash_bytes.copy_from_slice(&digest);
    let hash = B256::from(hash_bytes);
    let address = signature
        .recover_address_from_prehash(&hash)
        .map_err(|_| AuthorizationError::RecoveryFailed)?;
    Ok(address.into_array())
}
